src/traceme.c

```diff
@@ -158,71 +180,86 @@ write_buf_to_disk(int fd, void *buf, __u64 size) {
  */
-static void
+static int
 read_circular_buf(void *buf, __u64 size, __u64 head_monotonic, __u64 *tail_p, int out_fd) {
     __u64 head = head_monotonic % size; // Head must be manually wrapped.
     __u64 tail = *tail_p;
 
     if (tail <= head) {
         // No wrap-around
         DEBUG("read with no wrap");
-        write_buf_to_disk(out_fd, buf + tail, head - tail);
+        if (write_buf_to_disk(out_fd, buf + tail, head - tail) == -1) {
+            return -1;
+        }
     } else {
         // Wrap-around
         DEBUG("read with wrap");
-        write_buf_to_disk(out_fd, buf + tail, size - tail);
-        write_buf_to_disk(out_fd, buf, head);
+        if (write_buf_to_disk(out_fd, buf + tail, size - tail) == -1) {
+            return -1;
+        }
+        if (write_buf_to_disk(out_fd, buf, head) == -1) {
+            return -1;
+        }
     }
 
     // Update buffer tail, thus marking the space we just read as re-usable
     *tail_p = head;
+    return 0;
 }
 
 /*
  * Take trace data out of the AUX buffer.
+ *
+ * Returns 0 on success and -1 on error.
  */
-static void
-poll_loop(struct tracer_ctx *tr_ctx)
+static int
+poll_loop(int perf_fd, int stop_fd, int out_fd, struct perf_event_mmap_page *mmap_hdr, void *aux)
 {
-    struct perf_event_mmap_page *mmap_header = tr_ctx->mmap_header;
-    int n_events = 0;
+    int n_events = 0, ret = 0;
     size_t num_wakes = 0;
     struct pollfd pfds[2] = {
-        {tr_ctx->perf_fd,           POLLIN | POLLHUP,   0},
-        {tr_ctx->loop_done_fds[0],  POLLHUP,            0}
+        {perf_fd,   POLLIN | POLLHUP,   0},
+        {stop_fd,   POLLHUP,            0}
     };
 
     while (1) {
         n_events = poll(pfds, 2, INFTIM);
         if (n_events == -1) {
-            err(EXIT_FAILURE, "poll");
+            ret = -1;
+            goto done;
         }
 
         if ((pfds[0].revents & POLLIN) || (pfds[1].revents & POLLHUP)) {
             // See <linux/perf_event.h> for why we need the asm block.
-            __u64 head = mmap_header->aux_head;
+            __u64 head = mmap_hdr->aux_head;
             asm volatile ("" : : : "memory");
 
             // We were awoken to read out trace info, or we tracing stopped.
             num_wakes++;
             DEBUG("wake");
             DEBUG("aux_head=  0x%010llu", head);
-            DEBUG("aux_tail=  0x%010llu", mmap_header->aux_tail);
-            DEBUG("aux_offset=0x%010llu", mmap_header->aux_offset);
-            DEBUG("aux_size=  0x%010llu", mmap_header->aux_size);
-
-            read_circular_buf(tr_ctx->aux, mmap_header->aux_size,
-                head, &mmap_header->aux_tail, tr_ctx->out_fd);
+            DEBUG("aux_tail=  0x%010llu", mmap_hdr->aux_tail);
+            DEBUG("aux_offset=0x%010llu", mmap_hdr->aux_offset);
+            DEBUG("aux_size=  0x%010llu", mmap_hdr->aux_size);
+
+            if (read_circular_buf(aux, mmap_hdr->aux_size,
+                head, &mmap_hdr->aux_tail, out_fd) == -1) {
```

Comment:
The formatting of this call makes it very hard to read w.r.t. the next line.

Reply:
This is how it's done in KNF. How should I format it?

Reply:
I suggest we use the Rust idea: `head` would be aligned with `aux`.

Reply:
OK, so we really do need to draft up a C code style document.

Reply:
It would also be nice if we can find a vim config to do that alignment automatically.

Reply:
Good luck: I can't even get vim to reliably indent normal code properly!

Reply:
it's `:set cino+=(0`